osbuild/util/pe32p.py

```diff
@@ -0,0 +1,171 @@
+    "TimeDateStamp "
+    "PointerToSymbolTable "
+    "NumberOfSymbols "
+    "SizeOfOptionalHeader "
+    "Characteristics"
+)
+
+
+SectionFormat = "8sIIIIIIHHI"
+SectionHeader = namedtuple(
+    "SectionHeader",
+    "Name "
+    "VirtualSize "
+    "VirtualAddress "
+    "SizeOfRawData "
+    "PointerToRawData "
+    "PointerToRelocations "
+    "PointerToLinenumbers "
+    "NumberOfRelocations "
+    "NumberOfLinenumbers "
+    "Characteristics "
+)
+
+
+class SectionFlags(enum.Flag):
+    ALIGN_1BYTES = 0x00100000
+    ALIGN_2BYTES = 0x00200000
+    ALIGN_4BYTES = 0x00300000
+    ALIGN_8BYTES = 0x00400000
+    ALIGN_16BYTES = 0x00500000
+    ALIGN_32BYTES = 0x00600000
+    ALIGN_64BYTES = 0x00700000
+    ALIGN_128BYTES = 0x00800000
+    ALIGN_256BYTES = 0x00900000
+    ALIGN_512BYTES = 0x00A00000
+    ALIGN_1024BYTES = 0x00B00000
+    ALIGN_2048BYTES = 0x00C00000
+    ALIGN_4096BYTES = 0x00D00000
+    ALIGN_8192BYTES = 0x00E00000
+    ALIGN_MASK = 0x00F00000
+    ALIGN_DEFAULT = ALIGN_16BYTES
+
+
+OptionalFormat = "HBBIIIIIQIIHHHHHHIIIIHHQQQQII"
+OptionalHeader = namedtuple(
+    "OptionalHeader",
+    # Standard fields
+    "Magic "
+    "MajorLinkerVersion "
+    "MinorLinkerVersion "
+    "SizeOfCode "
+    "SizeOfInitializedData "
+    "SizeOfUninitializedData "
+    "AddressOfEntryPoint "
+    "BaseOfCode "
+    # Windows-Specific fields (PE32+)
+    "ImageBase "
+    "SectionAlignment "
+    "FileAlignment "
+    "MajorOperatingSystemVersion "
+    "MinorOperatingSystemVersion "
+    "MajorImageVersion "
+    "MinorImageVersion "
+    "MajorSubsystemVersion "
+    "MinorSubsystemVersion "
+    "Reserved1 "
+    "SizeOfImage "
+    "SizeOfHeaders "
+    "CheckSum "
+    "Subsystem "
+    "DllCharacteristics "
+    "SizeOfStackReserve "
+    "SizeOfStackCommit "
+    "SizeOfHeapReserve "
+    "SizeOfHeapCommit "
+    "LoaderFlags "
+    "NumberOfRvaAndSizes "
+)
+
+
+def seek_to_coff(f, check: bool = False):
+    if check:
+        f.seek(0, io.SEEK_SET)
+        magic = f.read(2)
+        assert magic == b"MZ", "Not a PE32+ file (missing DOS stub)"
+
+    f.seek(0x3c, io.SEEK_SET)
+    buf = f.read(struct.calcsize("I"))
+    (s, ) = struct.unpack_from("I", buf)
+    f.seek(int(s), io.SEEK_SET)
+
+
+def read_coff_header(f, check: bool = True) -> CoffHeader:
+    seek_to_coff(f, check=check)
+    buf = f.read(struct.calcsize(CoffFormat))
+    coff = CoffHeader._make(struct.unpack_from(CoffFormat, buf))
+    assert coff.Signature == b"PE\0\0", "Not a PE32+ file (missing PE header)"
+    return coff
+
+
+def read_optional_header(f, coff: Optional[CoffHeader] = None) -> OptionalHeader:
+    if coff is None:
+        coff = read_coff_header(f)
+
+    buf = f.read(coff.SizeOfOptionalHeader)
+    sz = struct.calcsize(OptionalFormat)
+    assert len(buf) >= sz, "Optional header too small"
+    opt = OptionalHeader._make(struct.unpack_from(OptionalFormat, buf))
+    assert opt.Magic == 0x20B, f"Not a PE32+ file (magic: {opt.Magic:X})"
```

Comment:
Is PE32+-only ok? 32bit PE binaries might not be encoded as PE32+, but they could. Lets add PE32 support when we encounter it?